File: akka-examples-actor/src/main/scala/com/github/tachesimazzoca/akka/example/heartbeat/Worker.scala
package com.github.tachesimazzoca.akka.example.heartbeat

import akka.actor.Actor
import akka.actor.ActorLogging

import scala.concurrent.duration._

object Worker {
  case object HeartBeat
  case class Alived(counter: Int)
}

class Worker extends Actor with ActorLogging {
  import context.dispatcher
  import Worker._

  var counter = 0

  def receive: Receive = alived

  val alived: Receive = {
    case HeartBeat =>
      counter += 1
      log.info(s"HeartBeat #$counter")
      if (counter == 5) {
        log.info("worker is now busy")
        context.become(busy)
        context.system.scheduler.scheduleOnce(10.seconds) {
          log.info("worker is now alived again after 10 seconds")
          context.become(alived)
        }
      }
      sender ! Alived(counter)
  }

  val busy: Receive = {
    case _ => // No reply
  }
}

File: java/akka-examples-reaper/src/main/java/com/github/tachesimazzoca/akka/examples/reaper/Worker.java
package com.github.tachesimazzoca.akka.examples.reaper;

import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import scala.concurrent.duration.Duration;

import java.util.Random;
import java.util.concurrent.TimeUnit;

public class Worker extends UntypedActor {
    private LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof Integer) {
            Integer n = (Integer) message;
            if (n > 0) {
                long sleepTime = 1000L * (new Random().nextInt(5) + 1);
                log.info(String.format("n: %d, sleepTime: %d", n, sleepTime));
                getContext().system().scheduler().scheduleOnce(
                        Duration.create(sleepTime, TimeUnit.MILLISECONDS),
                        getSelf(), n - 1, getContext().dispatcher(),
                        getSelf());
            } else {
                getContext().stop(getSelf());
            }

        } else {
            unhandled(message);
        }
    }
}
